crates/protocol/src/interop/test_utils.rs

```diff
@@ -0,0 +1,221 @@
+pub struct MockInteropProvider<T: TxReceipt<Log = Log>> {
+    /// A map of chain IDs to a map of block numbers to sealed headers.
+    pub headers: HashMap<u64, HashMap<u64, Sealed<Header>>>,
+    /// A map of chain IDs to a map of block numbers to receipts.
+    pub receipts: HashMap<u64, HashMap<u64, Vec<T>>>,
+}
+
+impl<T> MockInteropProvider<T>
+where
+    T: TxReceipt<Log = Log>,
+{
+    /// Construct a new mock interop provider.
+    pub const fn new(
+        headers: HashMap<u64, HashMap<u64, Sealed<Header>>>,
+        receipts: HashMap<u64, HashMap<u64, Vec<T>>>,
+    ) -> Self {
+        Self { headers, receipts }
+    }
+}
+
+#[async_trait]
+impl<T> InteropProvider for MockInteropProvider<T>
+where
+    T: TxReceipt<Log = Log>,
+{
+    type Receipt = T;
+
+    /// Fetch a [Header] by its number.
+    async fn header_by_number(&self, chain_id: u64, number: u64) -> InteropProviderResult<Header> {
+        Ok(self
+            .headers
+            .get(&chain_id)
+            .and_then(|headers| headers.get(&number))
+            .unwrap()
+            .inner()
+            .clone())
+    }
+
+    /// Fetch a [Header] by its hash.
+    async fn header_by_hash(&self, chain_id: u64, hash: B256) -> InteropProviderResult<Header> {
+        Ok(self
+            .headers
+            .get(&chain_id)
+            .and_then(|headers| headers.values().find(|header| header.hash() == hash))
+            .unwrap()
+            .inner()
+            .clone())
+    }
+
+    /// Fetch all receipts for a given block by number.
+    async fn receipts_by_number(
+        &self,
+        chain_id: u64,
+        number: u64,
+    ) -> InteropProviderResult<Vec<T>> {
+        Ok(self.receipts.get(&chain_id).and_then(|receipts| receipts.get(&number)).unwrap().clone())
+    }
+
+    /// Fetch all receipts for a given block by hash.
+    async fn receipts_by_hash(
+        &self,
+        chain_id: u64,
+        block_hash: B256,
+    ) -> InteropProviderResult<Vec<T>> {
+        Ok(self
+            .receipts
+            .get(&chain_id)
+            .and_then(|receipts| {
+                let headers = self.headers.get(&chain_id).unwrap();
+                let number =
+                    headers.values().find(|header| header.hash() == block_hash).unwrap().number;
+                receipts.get(&number)
+            })
+            .unwrap()
+            .clone())
+    }
+}
+
+/// A builder for constructing a superchain.
+#[derive(Debug)]
+pub struct SuperchainBuilder<T: TxReceipt<Log = Log> + alloc::fmt::Debug> {
+    /// Chains in the superchain
+    chains: HashMap<u64, ChainBuilder<T>>,
+    /// The current superchain timestamp.
+    timestamp: u64,
+}
+
+/// A builder for a single chain in a superchain.
+#[derive(Debug)]
+pub struct ChainBuilder<T: TxReceipt<Log = Log> + alloc::fmt::Debug> {
+    /// The header of the chain.
+    header: Header,
+    /// The receipts of the chain.
+    receipts: Vec<T>,
+}
+
+impl SuperchainBuilder<ReceiptEnvelope> {
```

Comment:
Shouldn't be hugely important, but notice that here we don't want to create a circular dependency between `op-alloy` and `mailli`, so instead of using the `OpReceiptEnvelope` defined in `op-alloy-consensus`, we use `ReceiptEnvelope` from `alloy-consensus`.

The test utilities themselves however have been made generic over the reveipt type such that kona may define this functionality for the `OpReceiptEnevelope` type.